### PR TITLE
Attach browser features to tracking requests when they are send

### PR DESCRIPTION
**Pull request by sgiehl**
### Description:

When performing tracking requests like below, it currently ends up in two visits:

```
        _paq.push(['requireCookieConsent']);
        _paq.push(['disableBrowserFeatureDetection']);
        _paq.push(['trackPageView']);
        _paq.push(['setCookieConsentGiven']);
        _paq.push(['trackPageView']);
```

The reason for this is, that the browser feature are detected with the first page view. The tracking request is then pushed to the consent queue. Once the consent is given, the browser feature detection is enabled automatically and the second tracking request will contain the now detected browser features.

Both requests will also contain different visitor ids, as it is generated using the current browser features.
As the config ids used in PHP to match the visitors is also based on the browser features, the both tracking requests end up in separate visits.

The changes in this PR, changes the behavior, so the browser features are attached to the tracking requests when they are sent out.
So both tracking requests will afterwards contain the same set of browser features. The visitor id, though will still differ, but on PHP side both actions will be mapped to the same visit, as the calculated config id matches.

Potential fix for #19279

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Review by bx80 — APPROVED** [review]
This appears to fix the problem detailed in the issue description and the relevant tests are passing :+1: 

* [x] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [x] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [x] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [x] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [x] n/a [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [x] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [x] Added [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [x] none [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [x] n/a [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [x] n/a [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [x] n/a Existing documentation updated if needed

**Comment by sgiehl** [comment]
@tsteur would you mind having a quick look at this PR as well? As it changes some parts of the tracking it might be good to have a second review on it to avoid possible issues.

File: js/piwik.js
Comment (tsteur) [inline js/piwik.js:3120]: I guess the `callback` addition is unrelated to the actual issue?

**Review by tsteur — COMMENTED** [review]
@sgiehl I think I still have two visits when following below steps:

```js
    var _paq = _paq || [];
    _paq.push(['enableLinkTracking']);
    _paq.push(['requireCookieConsent']);
    _paq.push(['trackPageView']);
```

Then editing the HTML and putting in:
```js
    var _paq = _paq || [];
    _paq.push(['enableLinkTracking']);
    _paq.push(['requireCookieConsent']);
    _paq.push(['trackPageView']);
    _paq.push(['setCookieConsentGiven']);
```

Then reloading the page.

I assume this is just not something we can prevent? Unless we always issued some `ping` or something when giving consent?

File: js/piwik.js
Comment (sgiehl) [reply js/piwik.js:3120]: I refactored that. The callbacks were before stored in a separate queue. I guess that should have worked. but in theory the could have been cases, where the consent and callback queue might get out of sync, causing the wrong callback being called.

**Comment by sgiehl** [comment]
@tsteur regarding your example: I tried to reproduce locally, but for me that ended up in only one visit.
Btw. when consent is given we already send a ping request if a tracking request was sent before:
https://github.com/matomo-org/matomo/blob/9b3a2511d5d804b0f204d1216ec1cf1dbc24d9d0/js/piwik.js#L6250-L6269

File: js/piwik.js
Comment (tsteur) [inline js/piwik.js:7332]: fyi this may also be triggered when the browser tab becomes hidden. not sure if this changes anything. Seems all fine though.

**Review by tsteur — COMMENTED** [review]
@sgiehl I can't reproduce this issue anymore. Tested quite a bit around regular consent and cookie consent and normal tracking code. Seems to work 👍 Nicely done

below the 2 visits I had yesterday but likely this was maybe due to some pre-existing cookies etc vs this time I made sure I had no cookies set initially etc 
image

File: js/piwik.js
Comment (sgiehl) [reply js/piwik.js:7332]: I've added that one more as kind of a safety net. Normally client hints should resolve within milliseconds. But as we can't know how client hints might be processed in the future there might be cases somewhen, where the client hints promise never resolves. Some browsers or plugins might e.g. possibly start blocking that fully or until the user confirms something... So guess it's fine to trigger that on tab unload as well.

**Comment by crankedguy** [comment]
@sgiehl : Hi, I am not sure if this is supposed to work already as #19279 was mentioned in the release log, because I can tell you that it definitely does not work in 5.0, at least not with MTM.
The follwing code has exactly the same behavior as before, meaning after giving consent browser features are enabled, tracking is correct with browser features and cookies, but there is a new visitor_id generated, as is a new visit. After withdrawing consent the tracking jumps back to the previous visitor without browser features and logs there. You can play this game endlessly back and forth.
```
<script type="text/javascript">
  var category = {{dk_consent}};
  var _paq = window._paq = window._paq || [];
  if ( category == "statistics" ) {
     console.log("We are logging from MTM with statistics, prio 99");
     _paq.push(['enableBrowserFeatureDetection']);
     _paq.push(['rememberCookieConsentGiven']);
  } else {
    console.log("We are logging from MTM without statistics, prio 99");
    _paq.push(['forgetCookieConsentGiven']);
  }
</script>
```

**Pull request merged**
